### ISSUE TITLE
Implement a streaming Identity Debugger

### ISSUE THREAD

**Issue by dabeeeenster** [SolidStateGroup/bullet-train-api#3459]
### Is your feature request related to a problem? Please describe.

Sometimes it can be hard to grok why identity data/sdk implementations are not working as expected

### Describe the solution you'd like.

Add a page in the dashboard that shows a real time stream of identities being sent to the API from SDKs. This is similar to real time debugger pages in analytics tools (https://www.docs.developers.amplitude.com/data/debugger/#ingestion-debugger)

### Describe alternatives you've considered

None

### Additional context

We could reuse the identity webhook to stream content back into the app.

**Comment by tranchung163** [comment]
I am working on this

**Comment by tranchung163** [comment]
Hi @dabeeeenster,

Our team is currently working on creating a graph page that will showcase the identity requests being sent from SDK clients to the REST API. We have a couple of questions and would appreciate your guidance:

Regarding the real-time data you mentioned, do you prefer implementing real-time streaming using WebSocket, or would a polling mechanism be sufficient for this purpose?

You mentioned an identity webhook in the documentation but we were unable to find specific details about it. Could you please point us in the right direction or provide more information on how to implement this webhook?

Additionally, we found the following two endpoints related to identity data:

api/v1/environments/{environment-api-key}/identities 
api/v1/environments

Are these the correct endpoints we should be using for fetching and streaming identity data? If there are any other relevant endpoints or additional considerations we should be aware of, please let us know.

Thank you for your assistance!

**Comment by matthewelwell** [comment]
Hi @tranchung163 , thanks for your work on this. 

In answer to your first question, I think polling is fine. 

Regarding the webhooks, the documentation is [here](https://docs.flagsmith.com/integrations/webhook). 

Regarding the endpoints, it gets a little more complicated when you talk about identities since there are 2 separate locations that the identity data can come from (for our SaaS platform at least). I've tried to detail this below, but please let me know if anything isn't clear. 

### Identities

#### Determining where to look for Identity data

The endpoint at `/api/v1/projects/:id` has an attribute called `use_edge_identities`. This value should be used to determine which endpoint(s) to use to retrieve data about a given environments identities. 

#### SaaS - Core & Self-Hosted (`"use_edge_identities": false`)

The endpoints at `api/v1/environments/{environment-api-key}/identities` will give you the data you need for the identities in this case. 

#### SaaS - Edge (`"use_edge_identities": true`)

The endpoints at `api/v1/environments/{environment-api-key}/edge-identities` will give you the data you need for the identities in this case. 

Note that we wouldn't necessarily expect you to do the work to handle the edge identities (although we are looking to open source our Edge API in the near future), but we may include it in our thinking when it comes to reviewing the implementation since certain usage patterns are more difficult when it comes to edge identities (since they are stored in DynamoDB vs Postgres.

**Comment by SebastianRamos3** [comment]
Hello,
Our team has been utilizing the /api/v1/environments/{environment-api-key}/identities endpoint, which we found to only return the count of identities and lacks real-time streaming capabilities.

To address this, we explored the /api/v1/organizations/{organization_pk}/usage-data endpoint and successfully created a usage data graph. However, this endpoint only provides identity tracking data at a daily granularity.
<img width="1440" alt="Screenshot 2024-08-01 at 3 49 02 PM" src="https://github.com/user-attachments/assets/bc7cf23d-a48f-4654-8410-760ef0a6028c">

To achieve more real-time streaming, we are considering modifying the database schema in api/app_analytics/influxdb_schema.py to track identity data at hourly and minute intervals.

Would implementing these changes align with the intended scope of the project, or is there an alternative approach recommended for real-time identity data streaming within Flagsmith?
Thank you for your assistance.

**Comment by dabeeeenster** [comment]
Can you clarify what services you are running? Are you using InfluxDB?

I'm also not 100% clear what you are trying to achieve? What is your use case? You interested in a real time stream of just historic data but with finer resolution?

**Comment by MinhTGit** [comment]
Hello @dabeeeenster, 

Sorry if we were not clear before. To answer your first question, yes, we are using InfluxDB and looking at api/app_analytics/influxdb_schema.py.

For the second question, our current graph shows the identities data at a daily level, and we want to display identities being sent by the hour or even minute. We believe this finer resolution would be more helpful for debugging.

In your response, you mentioned historic data. Could you clarify what you mean by this? Is there any other endpoint that allows us to get real-time stream identities? Are we heading in the right direction that you envision for this feature?

**Comment by dabeeeenster** [comment]
Can you make a PR and we can take a look at the code?

**Comment by SebastianRamos3** [comment]
We have gone ahead and made the pull request so you can take a look at the code.

**Comment by tranchung163** [comment]
Hi @dabeeeenster,
Thank you for your feedback, I’m trying to understand how to re-use the existing webhook infrastructure for this.
I understand that webhooks currently POST data in the following format: 
```python 
{
  "data": {
    "changed_by": "user@domain.com",
    "new_state": {
      "enabled": true,
      "environment": { "id": 23, "name": "Development" },
      "feature": { "id": 7168, "name": "butter_bar", "type": "CONFIG" },
      "feature_state_value": "<strong>You are using the develop environment.</strong>",
      "identity": null,
      "identity_identifier": null
    },
    "previous_state": { "enabled": false, ... },
    "timestamp": "2021-06-18T07:50:26.595298Z"
  },
  "event_type": "FLAG_UPDATED"
}
```
Is it possible to reuse the existing webhook POST payload like the one above to stream data to a new API endpoint (/api/v1/stream-identities) for example. If so, which specific data from the payload (like identity_name, email, timestamp, feature, etc.)? 

I’m exploring whether I can reuse the existing webhook handling mechanism from the file api/webhooks/webhooks.py, specifically the `_call_webhook` function, to push data to a new API endpoint for real-time streaming.
For example, I would like to add a section to the `_call_webhook` function like this to stream data back into new API endpoint  `(/api/v1/stream-identities)`:

```python
stream_res = requests.post("http://api/v1/stream-identities", data=json_data, headers=headers, timeout=10)
stream_res.raise_for_status()
```
Then add a page which shows identities data like identity_name, email, timestamp, feature, etc.? Does this seem like a feasible way to extend the webhook handling, or are there any potential pitfalls I should be aware of

Lastly, I’m unsure how to capture API calls made by the client-side SDK (e.g., `flagsmith.get_identity_flags(identifier="Development_user_123456")`) and then stream this identity call in real time. For instance, if a user requests `identity_flags.is_feature_enabled("my_cool_feature")`, do I need to capture this call for a specific identity, in this case it is "Development_user_123456"?

Thanks in advance for your help!

**Comment by matthewelwell** [comment]
@tranchung163 I'm not sure why you'd need to actually make any API request here? Couldn't you just pass the data directly into a separate service in the code (some kind of `identity_debugger_service` perhaps) without passing anything over the network?

I'd start by looking at the code in `api/integrations/integration.py` - can we add another element to the `IDENTITY_INTEGRATIONS` dictionary for example, that implements similar behaviour to the webhooks?

**Comment by tranchung163** [comment]
Hi @matthewelwell ,

I have found the `IDENTITY_INTEGRATION` and looked at the wrapper list in:
```python
IDENTITY_INTEGRATIONS: list[IntegrationConfig] = [
    {"relation_name": "amplitude_config", "wrapper": AmplitudeWrapper},
    {"relation_name": "heap_config", "wrapper": HeapWrapper},
    {"relation_name": "mixpanel_config", "wrapper": MixpanelWrapper},
    {"relation_name": "rudderstack_config", "wrapper": RudderstackWrapper},
    {"relation_name": "segment_config", "wrapper": SegmentWrapper},
    {"relation_name": "webhook_config", "wrapper": WebhookWrapper},
    {"relation_name": "identity_debugger_config", "wrapper": IdentityDebuggerWrapper}, # Add new identity_debugger element
]
```

I have added a new implementation for the [IdentityDebuggerWrapper]() with similar behavior to the webhooks:

```python
class IdentityDebuggerWrapper(AbstractBaseIdentityIntegrationWrapper):
    def __init__(self, config):
        self.config = config or {}

    def generate_user_data(self, identity, feature_states, trait_models=None):
        # Implement the logic to generate user data
        return {
            "identifier": identity.identifier,
            "feature_states": [
                {
                    "feature": feature_state.feature.name,
                    "enabled": feature_state.enabled,
                    "value": feature_state.get_feature_state_value(),
                }
                for feature_state in feature_states
            ],
            "traits": [
                {
                    "trait_key": trait.trait_key,
                    "trait_value": trait.trait_value,
                }
                for trait in (trait_models or [])
            ],
        }

    def identify_user_async(self, data):
        print("Identifying user with data:", data)
        processed_identities.append(data)
        # Broadcast the data to the WebSocket
        channel_layer = get_channel_layer()
        async_to_sync(channel_layer.group_send)(
            "identity_stream",
            {
                "type": "broadcast_identity",
                "identity": data,
            }
        )
```

My question is: 
Does this approach seem like the best way to solve the issue? 
If I want to stream data back to a new front-end webpage, can I get data from this implementation? If so, can I use WebSocket to achieve this?
Alternatively, can I re-use any element in the IDENTITY_INTEGRATION list to retrieve identity data from?

Looking forward to your help. Thank you in advance!

**Comment by matthewelwell** [comment]
> If I want to stream data back to a new front-end webpage, can I get data from this implementation? If so, can I use WebSocket to achieve this?

Yes, this seems like a reasonable approach. I've not used any streaming functionality in django before but [django-channels](https://channels.readthedocs.io/en/latest/) is probably a good place to start (although it might be overkill for our use case here)? 

> Alternatively, can I re-use any element in the IDENTITY_INTEGRATION list to retrieve identity data from?

I don't really understand what you mean here, perhaps you can elaborate?
